design/assets.unity.md

```diff
@@ -0,0 +1,157 @@
+* `assets` - The collection of assets generated from the given resource
+* `version` - The version of the loaded resource. Will typically be the HTTP response's `ETag` header.
+
+### LeaseAssets
+
+```cs
+Task<IEnumerable<UnityEngine.Object>> LeaseAssets(
+	string uri,
+	string ifMatches = null
+);
+```
+
+Asynchronously return the cached assets at the given URI, and increment the internal reference counter for this
+resource. Will return `null` if no assets are cached for that resource, or if `ifMatches` does not match the
+stored assets' `version`. This needs to be async in case the asset needs to be loaded from persistent storage.
+
+* `uri` - The resource identifier
+* `ifMatches` - If provided, assets will be returned only if the resource version matches this argument.
+
+### GetVersion
+
+```cs
+string GetVersion(string uri);
+```
+
+Returns the stored `version` of the given resource, or `null` if not cached. We'll need this for `If-Not-Match`
+HTTP headers.
+
+
+AssetManager
+---------------
+
+This class is responsible for maintaining references to active instances of assets used by a particular MRE instance.
+Every instance has its own AssetManager.
+
+### AssetMetadata
+
+```cs
+public struct AssetMetadata {
+	Guid Id;
+	Guid ContainerId;
+	UnityEngine.Object Asset;
+	ColliderGeometry ColliderGeometry;
+	AssetSource Source;
+}
+```
+
+Stores all the necessary info about an asset, including where it came from. `Source` will be null if this is not
+a shared asset, i.e. is a one-off creation of an MRE, or is a modified copy of something from the asset cache.
+
+### CacheRootGO
+
+The game object in the scene hierarchy that should be used as parent for any assets that require one, i.e. Prefabs.
+
+### EmptyTemplate
+
+The game object that should be duplicated for new actors.
+
+### GetById
+
+```cs
+AssetMetadata GetById(Guid? id, bool writeSafe = false);
+```
+
+Retrieve an asset by ID. If a write-safe asset is requested, and the stored asset with that ID is shared,
+a copy of the asset will be made, and stored back into the manager. Any other shared assets that reference this asset
+will also be recursively copied and stored back. Each copied asset will have the original returned to the cache,
+decrementing the original's reference count.
+
+### GetByObject
+
+```cs
+AssetMetadata GetByObject(UnityEngine.Object asset);
+```
+
+Retrieve an asset's metadata from the asset reference itself.
+
+### Set
+
+```cs
+void Set(AssetMetadata metadata);
+```
+
+Track a new asset reference. Will be called during asset creation, after the asset content is downloaded or retrieved
+from cache.
+
+### OnSet
+
+```cs
+void OnSet(Guid id, Action<AssetMetadata> callback);
+```
+
+Be notified when an asset is finished loading and available for use.
+
+### Unload
+
+```cs
+void Unload(Guid containerId);
+```
+
+Break references to all shared assets and destroy all unshared assets with this container ID.
+
+
+Expected flow: glTF
```

Comment:
I think this flow can be simplified a lot by:
* Disallowing the MRE instance to access the IAssetCache directly, and always go through its AssetManager.
* As I mentioned above, put IAssetCache in charge of fetching the asset from uri.

```C#
// Add an async method to AssetManager that will do the busywork of figuring out whether to look in the global
// cache, whether to copy for writeSafe, etc. Bonus if this can be a template method.
Task<AssetTypeT> AssetManager.GetAsset<AssetTypeT>(Uri uri, string ifVersionMatches = null, bool writeSafe = false);
```
```C#
// Usage example. This will fetch a newer version into the global cache if it was updated remotely.
var gltf = await this.assetManager.GetAsset<AssetTypeT>(uri, ifVersionMatches: null, writeSafe: true);
```

Reply:
Accessibility-wise, it's not possible to deny access to the MRE instance, given that the asset manager is *in* the instance, but that's a nitpick. I don't think it's worthwhile to pipe calls to the cache through the asset manager, since they don't share any functionality. The asset manager doesn't have the right information to fetch cached assets either (URL).